bindings/matrix-sdk-ffi/src/client.rs

```diff
@@ -258,6 +286,40 @@ impl Client {
             Ok(Arc::new(session_verification_controller))
         })
     }
+
+    /// Log out the current user
+    pub fn logout(&self) -> anyhow::Result<()> {
+        RUNTIME.block_on(async move {
+            _ = self.client.logout().await;
+            Ok(())
+        })
+    }
+
+    /// Process a sync error and return loop control accordingly
+    fn process_sync_error(&self, sync_error: Error) -> LoopCtrl {
+        let mut control = LoopCtrl::Continue;
+        if let Error::Http(HttpError::Api(FromHttpResponseError::Server(ServerError::Known(
+            RumaApiError::ClientApi(error),
+        )))) = sync_error
+        {
+            if matches!(error.kind, ErrorKind::UnknownToken { .. }) {
+                let is_soft_logout = match error.kind {
+                    ErrorKind::UnknownToken { soft_logout } => soft_logout,
+                    _ => unreachable!(),
+                };
+
+                self.state.write().unwrap().is_soft_logout = is_soft_logout;
+
+                if let Some(delegate) = &*self.delegate.read().unwrap() {
+                    delegate.did_update_restore_token();
+                    delegate.did_receive_auth_error(is_soft_logout);
+                }
+
+                control = LoopCtrl::Break
+            }
```

Comment:
Let's reduce this to:

```suggestion
            if let ErrorKind::UnknownToken { soft_logout } = error.kind {
                self.state.write().unwrap().is_soft_logout = soft_logout;

                if let Some(delegate) = &*self.delegate.read().unwrap() {
                    delegate.did_update_restore_token();
                    delegate.did_receive_auth_error(is_soft);
                }

                control = LoopCtrl::Break
            }
```